[RISCV] Handle fixed length vectors with exact VLEN in lowerINSERT_SUBVECTOR

lukel97 opened the pull request:
This is the insert_subvector equivalent to #79949, where we can avoid sliding up by the full LMUL amount if we know the exact subregister the subvector will be inserted into.

This mirrors the lowerEXTRACT_SUBVECTOR changes in that we handle this in two parts:

- We handle fixed length subvector types by converting the subvector to a scalable vector. But unlike EXTRACT_SUBVECTOR, we may also need to convert the vector being inserted into too.

- Whenever we don't need a vslideup because either the subvector fits exactly into a vector register group *or* the vector is undef, we need to emit an insert_subreg ourselves because RISCVISelDAGToDAG::Select doesn't correctly handle fixed length subvectors yet: see d7a28f7ad

A subvector exactly fits into a vector register group if its size is a known multiple of the size of a vector register, and this adds a new overload for TypeSize::isKnownMultipleOf for scalable to scalable comparisons to help reason about this.

I've left RISCVISelDAGToDAG::Select untouched for now (minus relaxing an invariant), so that the insert_subvector and extract_subvector code paths are the same.

We should teach it to properly handle fixed length subvectors in a follow-up patch, so that the "exact subregsiter" logic is handled in one place instead of being spread across both RISCVISelDAGToDAG.cpp and RISCVISelLowering.cpp.

lukel97 committed:
[RISCV] Handle fixed length vectors with exact VLEN in lowerINSERT_SUBVECTOR

This is the insert_subvector equivalent to #79949, where we can avoid sliding up
by the full LMUL amount if we know the exact subregister the subvector will be
inserted into.

This mirrors the lowerEXTRACT_SUBVECTOR changes in that we handle this in two
parts:

- We handle fixed length subvector types by converting the subvector to a
  scalable vector. But unlike EXTRACT_SUBVECTOR, we may also need to convert the
  vector being inserted into too.

- Whenever we don't need a vslideup because either the subvector aligns to a
  vector register group *or* the vector is undef, we need to emit an
  insert_subreg ourselves because RISCVISelDAGToDAG::Select doesn't correctly
  handle fixed length subvectors yet: see d7a28f7ad

I've left RISCVISelDAGToDAG::Select untouched for now (minus relaxing an
invariant), so that the insert_subvector and extract_subvector code paths are
the same.

We should teach it to properly handle fixed length subvectors in a follow-up
patch, so that the "exact subregsiter" logic is handled in one place instead of
being spread across both RISCVISelDAGToDAG.cpp and RISCVISelLowering.cpp.

preames commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -2063,8 +2063,14 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
     if (SubVecVT.isFixedLengthVector()) {
-      assert(Idx == 0 && V.isUndef());
       SubVecContainerVT = TLI.getContainerForFixedLengthVector(SubVecVT);
+      bool AlignedToVecReg = false;
```
Two points:
1) Naming - This doesn't appear to be checking alignment, it appears to be checking whether the fixed vector completely fills the container.
2) Variable used under assert only should be in appropriate ifndef block.

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9596,6 +9596,21 @@ SDValue RISCVTargetLowering::lowerVPREDUCE(SDValue Op,
+  if (auto VLen = Subtarget.getRealVLen()) {
+    const unsigned Vscale = *VLen / RISCV::RVVBitsPerBlock;
+    if (LHS.isScalable())
+      LHS = ElementCount::getFixed(LHS.getKnownMinValue() * Vscale);
```
You've got this pattern repeating a few times, maybe extract a helper which returns the ElementCount of a vector which returns a fixed value if possible given exact vlen?

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9698,41 +9714,92 @@ SDValue RISCVTargetLowering::lowerINSERT_SUBVECTOR(SDValue Op,
   bool IsSubVecPartReg = SubVecLMUL == RISCVII::VLMUL::LMUL_F2 ||
                          SubVecLMUL == RISCVII::VLMUL::LMUL_F4 ||
                          SubVecLMUL == RISCVII::VLMUL::LMUL_F8;
+  bool AlignedToVecReg = !IsSubVecPartReg;
```
Again, not alignment in the naming here, and another instance of getting a fixed size from a scalable quantity.

lukel97 replied:
I've gone a step further here to try and simplify this. We can check if the SubVec is known to completely fill the container if its size is a multiple of a vector register.

We already have a `isKnownMultipleOf` method on TypeSize that works on scalar quantities, but I've added a new variant that also works on scalable quantities. 

This way we can unify the fixed and scalable reasoning.

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -9698,41 +9714,92 @@ SDValue RISCVTargetLowering::lowerINSERT_SUBVECTOR(SDValue Op,
-  // lowering to an EXTRACT_SUBVECTOR grabbing the nearest LMUL=1 vector type
-  // (which resolves to a subregister copy), performing a VSLIDEUP to place the
-  // subvector within the vector register, and an INSERT_SUBVECTOR of that
+  // 2. If the subvector isn't exactly aligned to a vector register group, then
```
I think you actually want something along lines of "is an exact multiple of a valid register group size".

preames reviewed (commented):
This looks reasonable likely to be correct.  I've added a couple small comments on style to make things easier to follow, but I don't see any major problems.

Note to future self - I've glanced at code only.  Need to go through the test diffs carefully to see if that raises any questions.

lukel97 committed:
Address review comments:

* Rework the subvec size check by introducing a isKnownMultipleOf helper function on TypeSize/ElementCount. We can then reason that the subvector is exactly vector register sized if it is a multiple of the vector register size.
* Rename variables to clarify we are checking the subvec size
* Update wording of comment

lukel97 committed:
Add asserts that the subvector size is a power of 2

This is an invariant needed if we want to check that the subvector exactly fills a vector register by checking if it's a multiple of vlen.

lukel97 commented:
Ping

lukel97 committed:
Add [[maybe_unused]], remove unused check prefix

An unused check prefix apparently causes an error? Not sure if it was
always like that.

topperc commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -2063,8 +2063,16 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
       SubVecContainerVT = TLI.getContainerForFixedLengthVector(SubVecVT);
+#ifndef NDEBUG
+      TypeSize VecRegSize = TypeSize::getScalable(RISCV::RVVBitsPerBlock);
+      bool ExactlyVecRegSized =
```
Doesn't the `#ifndef NDEBUG` make the [[maybe_unused]] unnecessary?

lukel97 replied:
We probably want to actually run these assertions on release builds with assertions enabled, so I'll remove the NDEBUG check

preames reviewed (approved):
LGTM

lukel97 committed:
Remove NDEBUG gating around assertions
